File: ui/src/types/tiering.ts
// Mirrors be/internal/types/tiering.go — keep field names/shape in sync with
// the Go json tags (the admin tiering-report/apply contract).

export interface TieringDefRow {
  workflow_id: string
  def_id: string
  role: string
  current_model: string
  current_effort?: string
  recommended_model: string
  recommended_effort?: string
  recommended_template: string
  customized: boolean
  skip_reason?: string
  est_monthly_delta: number | null
  is_worker: boolean
}

export interface TieringProjectReport {
  project_id: string
  project_name: string
  defs: TieringDefRow[]
  est_monthly_delta: number | null
}

export interface TieringReport {
  projects: TieringProjectReport[]
  markdown: string
}

export interface TieringDefKey {
  workflow_id: string
  def_id: string
}

export interface TieringApplyConfirmation {
  project_id: string
  def_keys?: TieringDefKey[]
  confirm_all?: boolean
}

export interface TieringApplyRequest {
  confirmations: TieringApplyConfirmation[]
}

export interface TieringApplyOutcome {
  project_id: string
  workflow_id: string
  def_id: string
  role: string
  outcome: string
  reason?: string
}

export interface TieringApplyResult {
  applied: TieringApplyOutcome[]
  skipped: TieringApplyOutcome[]
}

File: be/internal/types/tiering.go
package types

// TieringDefRow is one agent_definitions row's current-vs-recommended tier
// state within a project's tiering report. CurrentModel/CurrentEffort and
// RecommendedModel/RecommendedEffort are CHAIN-RESOLVED effective values
// (the def's own tier-chain primary, or the recommended tier's chain
// primary) — not the raw agent_definitions.model/reasoning_effort columns,
// which are empty for a tier-driven def.
type TieringDefRow struct {
	WorkflowID          string   `json:"workflow_id"`
	DefID               string   `json:"def_id"`
	Role                string   `json:"role"`
	CurrentTier         *int     `json:"current_tier"`
	CurrentModel        string   `json:"current_model"`
	CurrentEffort       string   `json:"current_effort,omitempty"`
	RecommendedTier     int      `json:"recommended_tier"`
	RecommendedModel    string   `json:"recommended_model"`
	RecommendedEffort   string   `json:"recommended_effort,omitempty"`
	RecommendedTemplate string   `json:"recommended_template"`
	Customized          bool     `json:"customized"`
	SkipReason          string   `json:"skip_reason,omitempty"`
	EstMonthlyDelta     *float64 `json:"est_monthly_delta,omitempty"`
	IsWorker            bool     `json:"is_worker"`
	GrantsDelegation    bool     `json:"grants_delegation"`
}

// TieringProjectReport is one project's slice of the tiering report.
type TieringProjectReport struct {
	ProjectID       string          `json:"project_id"`
	ProjectName     string          `json:"project_name"`
	Defs            []TieringDefRow `json:"defs"`
	EstMonthlyDelta *float64        `json:"est_monthly_delta,omitempty"`
}

// TieringReport is the full dry-run response for GET /api/v1/admin/tiering-report.
type TieringReport struct {
	Projects []TieringProjectReport `json:"projects"`
	Markdown string                 `json:"markdown"`
}

// TieringDefKey identifies one agent_definitions row for apply confirmation.
type TieringDefKey struct {
	WorkflowID string `json:"workflow_id"`
	DefID      string `json:"def_id"`
}

// TieringApplyConfirmation is one project's explicit confirmation for
// POST /api/v1/admin/tiering-apply: either an explicit def key list, or
// ConfirmAll to apply every eligible def in the project.
type TieringApplyConfirmation struct {
	ProjectID  string          `json:"project_id"`
	DefKeys    []TieringDefKey `json:"def_keys,omitempty"`
	ConfirmAll bool            `json:"confirm_all,omitempty"`
}

// TieringApplyRequest is the POST /api/v1/admin/tiering-apply request body.
type TieringApplyRequest struct {
	Confirmations []TieringApplyConfirmation `json:"confirmations"`
}

// TieringApplyOutcome is the per-def result of an apply attempt.
type TieringApplyOutcome struct {
	ProjectID  string `json:"project_id"`
	WorkflowID string `json:"workflow_id"`
	DefID      string `json:"def_id"`
	Role       string `json:"role"`
	Outcome    string `json:"outcome"` // applied|unchanged|skipped-customized|skipped-consultant|skipped-hotfix|skipped-non-static|skipped-unconfirmed
	Reason     string `json:"reason,omitempty"`
}

// TieringApplyResult is the POST /api/v1/admin/tiering-apply response.
type TieringApplyResult struct {
	Applied []TieringApplyOutcome `json:"applied"`
	Skipped []TieringApplyOutcome `json:"skipped"`
}
